File: game_state.py
# -*- coding: utf-8 -*-
"""
SAO Auto — 统一游戏状态模型

所有识别结果汇总到 GameState，UI 只订阅此对象。

"""

import time
import threading
from dataclasses import dataclass, field
from typing import Optional, List, Callable

from perf_probe import probe as _probe

# 缓存字段名列表
_CACHE_FIELDS = (
    'player_name', 'level_base', 'level_extra', 'player_id',
    'season_exp', 'fight_point',
    'hp_current', 'hp_max', 'hp_pct',
    'stamina_current', 'stamina_max', 'stamina_pct',
    'profession_id', 'profession_name',
)

# 身份类字段: 仅当新值有效 (非零/非空) 时才覆写缓存
_CACHE_IDENTITY_FIELDS = frozenset((
    'player_name', 'level_base', 'level_extra', 'player_id',
    'profession_id', 'profession_name', 'fight_point',
))


def compute_burst_ready(skill_slots, watched_slots) -> bool:
    """Return True when every watched slot is ready and at least one matched."""
    if not skill_slots or not watched_slots:
        return False
    try:
        watched = {int(x) for x in watched_slots if int(x) > 0}
    except Exception:
        watched = set()
    if not watched:
        return False

    matched = []
    for slot in skill_slots:
        if not isinstance(slot, dict):
            continue
        try:
            idx = int(slot.get('index', 0) or 0)
        except Exception:
            idx = 0
        if idx in watched:
            matched.append(slot)
    if not matched:
        return False

    def _slot_ready(slot):
        state = str(slot.get('state', '') or '').strip().lower()
        if state:
            if state in ('ready', 'active'):
                return True
        try:
            if bool(slot.get('active')):
                return True
        except Exception:
            pass
        try:
            charge_count = int(slot.get('charge_count', 0) or 0)
        except Exception:
            charge_count = 0
        if charge_count > 0:
            return True
        try:
            remaining_ms = int(slot.get('remaining_ms', 0) or 0)
            if remaining_ms <= 120:
                return True
        except Exception:
            pass
        value = slot.get('cooldown_pct', 1.0)
        try:
            return float(value) <= 0.02
        except Exception:
            return False

    return all(_slot_ready(slot) for slot in matched)


@dataclass
class GameState:
    """游戏状态快照 — 所有字段由识别层写入，UI 层只读。"""

    # ── 身份信息 ──
    player_name: str = ''
    level_base: int = 0                # 基础等级 (0=未识别)
    level_extra: int = 0               # 括号内加成等级 (+XX)
    season_exp: int = 0                # 当前赛季/extraLevel EXP
    player_id: str = ''                # 玩家编号
    fight_point: int = 0               # 战斗力

    # ── 生命值 ──
    hp_current: int = 0
    hp_max: int = 0
    hp_pct: float = 1.0                # 0.0 ~ 1.0

    # ── 体力值 ──
    stamina_current: int = 0
    stamina_max: int = 0
    stamina_pct: float = 1.0
    stamina_offline: bool = False          # True when STA bar not visible

    # ── 技能栏 ──
    skill_slots: List = field(default_factory=list)
    # 每个元素:
    # {
    #   'index': int,
    #   'rect': {'x': int, 'y': int, 'w': int, 'h': int},
    #   'state': 'ready' | 'cooldown' | 'insufficient_energy' | 'unknown',
    #   'cooldown_pct': float,
    #   'insufficient_energy': bool,
    #   'active': bool,
    #   'ready_edge': bool,
    # }

    # ── Burst Mode Ready (CD 提醒) ──
    burst_ready: bool = False          # 所有监视技能 CD 就绪时为 True

    profession_id: int = 0             # 职业 ID (来自抓包)
    profession_name: str = ''          # 职业名称

    # ── 战斗状态 ──
    in_combat: bool = False            # AttrCombatState (104) — 是否在战斗中

    # ── 窗口信息 ──
    window_rect: Optional[tuple] = None    # (left, top, right, bottom)
    window_width: int = 0
    window_height: int = 0

    # ── Boss Raid ──
    boss_raid_active: bool = False
    boss_raid_phase: int = 0
    boss_raid_phase_name: str = ''
    boss_enrage_remaining: float = 0.0     # seconds till enrage (0=inactive)
    boss_timer_text: str = ''              # formatted text for ID plate
    boss_total_damage: int = 0
    boss_dps: int = 0
    boss_hp_est_pct: float = 1.0           # estimated boss HP% (1.0=full)
    boss_current_hp: int = 0               # real HP from packets
    boss_total_hp: int = 0                 # real MaxHP from packets (or profile)
    boss_hp_source: str = 'none'           # 'packet' | 'estimate' | 'none'
    boss_shield_active: bool = False
    boss_shield_pct: float = 0.0
    boss_breaking_stage: int = -1          # -1=not received; 0=Breaking; 1=BreakEnd
    boss_extinction_pct: float = 0.0
    boss_in_overdrive: bool = False
    boss_invincible: bool = False

    # ── 采集元信息 ──
    capture_timestamp: float = 0.0
    recognition_ok: bool = False
    packet_active: bool = False
    error_msg: str = ''
    identity_alert_serial: int = 0
    identity_alert_title: str = ''
    identity_alert_message: str = ''

    @property
    def level_text(self) -> str:
        """格式化等级文本: 60(+12)"""
        if self.level_extra > 0:
            return f'{self.level_base}(+{self.level_extra})'
        return str(self.level_base)

    @property
    def hp_text(self) -> str:
        return f'{self.hp_current}/{self.hp_max}'

    @property
    def stamina_text(self) -> str:
        return f'{int(round(max(0.0, min(1.0, self.stamina_pct)) * 100.0))}%'

    def to_dict(self) -> dict:
        return {
            'player_name': self.player_name,
            'level_base': self.level_base,
            'level_extra': self.level_extra,
            'season_exp': self.season_exp,
            'level_text': self.level_text,
            'player_id': self.player_id,
            'hp_current': self.hp_current,
            'hp_max': self.hp_max,
            'hp_pct': round(self.hp_pct, 4),
            'stamina_current': self.stamina_current,
            'stamina_max': self.stamina_max,
            'stamina_pct': round(self.stamina_pct, 4),
            'skill_slots': list(self.skill_slots),
            'burst_ready': bool(self.burst_ready),
            'profession_id': self.profession_id,
            'profession_name': self.profession_name,
            'hp_text': self.hp_text,
            'stamina_text': self.stamina_text,
            'recognition_ok': self.recognition_ok,
            'packet_active': self.packet_active,
            'capture_ts': self.capture_timestamp,
            'boss_raid_active': self.boss_raid_active,
            'boss_raid_phase': self.boss_raid_phase,
            'boss_raid_phase_name': self.boss_raid_phase_name,
            'boss_enrage_remaining': round(self.boss_enrage_remaining, 1),
            'boss_timer_text': self.boss_timer_text,
            'boss_total_damage': self.boss_total_damage,
            'boss_dps': self.boss_dps,
            'boss_hp_est_pct': round(self.boss_hp_est_pct, 4),
            'boss_current_hp': self.boss_current_hp,
            'boss_total_hp': self.boss_total_hp,
            'boss_hp_source': self.boss_hp_source,
            'boss_shield_active': self.boss_shield_active,
            'boss_shield_pct': round(self.boss_shield_pct, 4),
            'boss_breaking_stage': self.boss_breaking_stage,
            'boss_extinction_pct': round(self.boss_extinction_pct, 4),
            'boss_in_overdrive': self.boss_in_overdrive,
            'boss_invincible': self.boss_invincible,
            'identity_alert_serial': self.identity_alert_serial,
            'identity_alert_title': self.identity_alert_title,
            'identity_alert_message': self.identity_alert_message,
        }


class GameStateManager:
    """线程安全的状态管理器，支持订阅通知。"""

    def __init__(self):
        self._state = GameState()
        self._lock = threading.Lock()
        self._listeners: List[Callable[[GameState], None]] = []

    @property
    def state(self) -> GameState:
        with self._lock:
            return self._state

    @_probe.decorate('state.update')
    def update(self, **kwargs):
        """部分更新状态字段并通知所有监听器 (含范围校验)"""
        with self._lock:
            # ── 预过滤: 拦截 HP/LV/STA 的 0 值, 在 setattr 之前保护 ──
            allow_zero_hp = False
            if 'hp_pct' in kwargs:
                try:
                    allow_zero_hp = float(kwargs['hp_pct']) <= 0.001
                except Exception:
                    allow_zero_hp = False
            if 'hp_current' in kwargs and kwargs['hp_current'] == 0:
                if self._state.hp_max > 0 and not allow_zero_hp:
                    prev = getattr(self, '_prev_hp_current', self._state.hp_current)
                    if prev > 0:
                        kwargs['hp_current'] = prev
                    else:
                        kwargs['hp_current'] = self._state.hp_max
            if 'level_base' in kwargs and kwargs['level_base'] == 0:
                prev_lv = getattr(self, '_prev_level_base', self._state.level_base)
                if prev_lv > 0:
                    kwargs['level_base'] = prev_lv
            if 'level_extra' in kwargs and kwargs['level_extra'] == 0:
                prev_lv_extra = getattr(self, '_prev_level_extra', self._state.level_extra)
                if prev_lv_extra > 0:
                    kwargs['level_extra'] = prev_lv_extra
            if 'stamina_current' in kwargs and kwargs['stamina_current'] == 0:
                next_sta_max = kwargs.get('stamina_max', self._state.stamina_max)
                if 'stamina_pct' not in kwargs and int(next_sta_max or 0) <= 0:
                    prev_sta = getattr(self, '_prev_stamina_current', self._state.stamina_current)
                    if prev_sta > 0:
                        kwargs['stamina_current'] = prev_sta

            for k, v in kwargs.items():
                if not hasattr(self._state, k):
                    continue
                # ── 范围校验 ──
                if k in ('hp_pct', 'stamina_pct'):
                    if not isinstance(v, (int, float)):
                        continue
                    v = max(0.0, min(1.0, float(v)))
                elif k == 'level_base':
                    if not isinstance(v, int) or v < 0 or v > 999:
                        continue
                elif k == 'level_extra':
                    if not isinstance(v, int) or v < 0 or v > 999:
                        continue
                elif k == 'season_exp':
                    if not isinstance(v, int) or v < 0:
                        continue
                elif k in ('hp_current', 'hp_max', 'stamina_current', 'stamina_max'):
                    if not isinstance(v, int) or v < 0:
                        continue
                elif k == 'player_name':
                    if not isinstance(v, str) or len(v) > 20:
                        continue
                elif k == 'player_id':
                    if not isinstance(v, str) or len(v) > 30:
                        continue
                elif k == 'identity_alert_serial':
                    if not isinstance(v, int) or v < 0:
                        continue
                elif k == 'identity_alert_title':
                    if not isinstance(v, str) or len(v) > 80:
                        continue
                elif k == 'identity_alert_message':
                    if not isinstance(v, str) or len(v) > 600:
                        continue
                elif k == 'boss_raid_phase_name':
                    if not isinstance(v, str) or len(v) > 60:
                        continue
                elif k == 'boss_enrage_remaining':
                    if not isinstance(v, (int, float)) or v < 0:
                        continue
                    v = float(v)
                elif k == 'boss_timer_text':
                    if not isinstance(v, str) or len(v) > 40:
                        continue
                elif k in ('boss_total_damage', 'boss_dps'):
                    if not isinstance(v, int) or v < 0:
                        continue
                elif k == 'boss_hp_est_pct':
                    if not isinstance(v, (int, float)):
                        continue
                    v = max(0.0, min(1.0, float(v)))
                setattr(self._state, k, v)

            # ── 更新 prev 追踪值 (仅当值有效时) ──
            if self._state.hp_current > 0:
                self._prev_hp_current = self._state.hp_current
            if self._state.level_base > 0:
                self._prev_level_base = self._state.level_base
            if self._state.level_extra > 0:
                self._prev_level_extra = self._state.level_extra
            if self._state.stamina_current > 0:
                self._prev_stamina_current = self._state.stamina_current

            # hp_current > hp_max: 回滚 current
            if self._state.hp_max > 0 and self._state.hp_current > self._state.hp_max:
                self._state.hp_current = self._state.hp_max
            if self._state.stamina_max > 0 and self._state.stamina_current > self._state.stamina_max:
                self._state.stamina_current = self._state.stamina_max
            self._state.capture_timestamp = time.time()
            snapshot = GameState(**{
                f.name: getattr(self._state, f.name)
                for f in self._state.__dataclass_fields__.values()
            })
        # 通知在锁外执行，避免死锁
        for cb in self._listeners:
            try:
                cb(snapshot)
            except Exception as e:
                print(f'[GameState] listener error: {e}')

    def subscribe(self, callback: Callable[[GameState], None]):
        self._listeners.append(callback)

    def unsubscribe(self, callback: Callable[[GameState], None]):
        try:
            self._listeners.remove(callback)
        except ValueError:
            pass

    # ── 缓存持久化 ──

    def load_cache(self, settings):
        """从 settings 加载上次缓存的游戏状态 (启动时调用)"""
        cache = settings.get('game_cache', {})
        if not cache or not isinstance(cache, dict):
            return

        with self._lock:
            for k in _CACHE_FIELDS:
                v = cache.get(k)
                if v is not None and hasattr(self._state, k):
                    try:
                        expected_type = type(getattr(self._state, k))
                        setattr(self._state, k, expected_type(v))
                    except (ValueError, TypeError):
                        pass
            print(f'[GameState] 从缓存加载: HP={self._state.hp_current}/{self._state.hp_max}, '
                  f'LV={self._state.level_base}'
                  f'{"(+" + str(self._state.level_extra) + ")" if self._state.level_extra > 0 else ""}')
        # 通知订阅者立即渲染缓存数据 (避免等待首个数据包)
        try:
            snapshot = GameState(**{
                f.name: getattr(self._state, f.name)
                for f in self._state.__dataclass_fields__.values()
            })
            for cb in self._listeners:
                try:
                    cb(snapshot)
                except Exception:
                    pass
        except Exception:
            pass

    def save_cache(self, settings):
        """将当前状态持久化到 settings (定期调用)

        身份类字段 (name/level/profession) 仅在收到有效值时才覆写缓存,
        避免工具中途启动时用默认 0/空字符串覆盖上次缓存的值。
        """
        with self._lock:
            # 以现有缓存为基底, 避免丢失上次保存的身份数据
            cache = dict(settings.get('game_cache', {}) or {})
            for k in _CACHE_FIELDS:
                v = getattr(self._state, k, None)
                if v is None:
                    continue
                # 身份类字段: 仅当值非零/非空时覆写
                if k in _CACHE_IDENTITY_FIELDS:
                    if isinstance(v, str) and not v.strip():
                        continue
                    if isinstance(v, (int, float)) and v <= 0:
                        continue
                cache[k] = v
        settings.set('game_cache', cache)
        settings.save()

File: C#/src/SaoAuto.Core/State/GameState.cs
using System.Collections.Immutable;
using SaoAuto.Core.Packets;

namespace SaoAuto.Core.State;

/// <summary>
/// Immutable snapshot of the unified game state.
/// Mirrors fields of Python <c>game_state.GameState</c>.
/// UI/overlay code must treat instances as read-only.
/// </summary>
public sealed record GameState
{
    // ── Identity ──
    public string PlayerName { get; init; } = string.Empty;
    public int LevelBase { get; init; }
    public int LevelExtra { get; init; }
    public int SeasonExp { get; init; }
    public string PlayerId { get; init; } = string.Empty;
    public int FightPoint { get; init; }

    // ── HP ──
    public int HpCurrent { get; init; }
    public int HpMax { get; init; }
    public double HpPct { get; init; } = 1.0;

    // ── Stamina ──
    public int StaminaCurrent { get; init; }
    public int StaminaMax { get; init; }
    public double StaminaPct { get; init; } = 1.0;
    public bool StaminaOffline { get; init; }

    // ── Skill bar / burst ──
    public ImmutableArray<SkillSlot> SkillSlots { get; init; } = ImmutableArray<SkillSlot>.Empty;
    public bool BurstReady { get; init; }
    public int ProfessionId { get; init; }
    public string ProfessionName { get; init; } = string.Empty;

    // ── Buff monitor ──
    public ImmutableArray<BuffEntry> SelfBuffs { get; init; } = ImmutableArray<BuffEntry>.Empty;
    public double ServerTimeOffsetMs { get; init; }

    // ── Combat ──
    public bool InCombat { get; init; }

    // ── Self identity (S64: from EnterGame) ──
    public ulong SelfUuid { get; init; }

    // ── Skill cooldown table (S64: from SyncClientUseSkill) ──
    // Keyed by SkillLevelId; value = timestampSeconds of last use.
    public ImmutableDictionary<int, double> SkillLastUseAt { get; init; }
        = ImmutableDictionary<int, double>.Empty;

    // ── Server-reported skill cooldowns (S69: from SyncToMeDeltaInfo) ──
    // Keyed by SkillLevelId; value = full per-CD snapshot (begin/dur/vcd/
    // charge/sub/accel). Populated by AoiSyncToMeDelta.SyncSkillCDs;
    // expired entries are pruned on each delta.
    public ImmutableDictionary<int, SkillCdSnapshot> SkillCdMap { get; init; }
        = ImmutableDictionary<int, SkillCdSnapshot>.Empty;

    // ── Server-reported fight-resource cooldowns (S108: from
    // SyncToMeDeltaInfo.FightResCDs sibling event). Keyed by ResId;
    // value = (begin/dur/vcd) per-resource snapshot. Populated by
    // ApplyToMeFightResCd; expired entries pruned on each delta.
    // Mirrors Python's fight_res_cd_map at packet_parser.py 3967–3973.
    public ImmutableDictionary<int, FightResCdSnapshot> FightResCdMap { get; init; }
        = ImmutableDictionary<int, FightResCdSnapshot>.Empty;

    // ── Buff-driven CD modifier scalars (S122: from
    // SyncToMeDeltaInfo BaseDelta.TempAttrs). Cumulative sums of
    // TempAttr ids 100/101/103 over the collection — pct + accel are
    // /10000 万分比 ratios; fixed is in ms. Self-only; bridge filters
    // by SelfUuid. Consumers that care about server-side CDR
    // (skill HUD timer compression) read these alongside SkillCdMap.
    // Mirrors Python's `player.temp_attr_cd_*` fields written by
    // `_process_temp_attr_collection` at packet_parser.py 4831–4869.
    public int TempAttrCdPct { get; init; }
    public int TempAttrCdFixed { get; init; }
    public int TempAttrCdAccel { get; init; }

    // ── Dungeon (S64: from SyncDungeonData / SyncDungeonDirtyData) ──
    public long DungeonSceneUuid { get; init; }
    public int DungeonDifficulty { get; init; }
    public int? DungeonFlowState { get; init; }
    public ImmutableArray<DungeonTargetProgress> DungeonTargets { get; init; }
        = ImmutableArray<DungeonTargetProgress>.Empty;

    // ── Near-entity table (S68: from SyncNearEntities) ──
    // Keyed by entity Uuid; value = (EntityType, FirstSeenSeconds).
    // Removed when SyncNearEntities reports a Disappear for the uuid.
    public ImmutableDictionary<long, EntityTableEntry> NearEntities { get; init; }
        = ImmutableDictionary<long, EntityTableEntry>.Empty;

    // ── Monster data table (S113: skeleton populated from
    // SyncNearEntities + SyncNearDeltaInfo). Keyed by uuid; value =
    // a record with the per-monster fields the entity table can't
    // hold (Name / TemplateId / breaking gauge / shield / IsDead).
    // Mirrors Python's `MonsterData` at packet_parser.py 803. Lazy-
    // create on first appearance for entities with EntityType==1
    // (Entmonster); removed on Disappear; HP slot mirrors the entity
    // table so death (CurHp=0 with HasCurHp=true) flips IsDead. The
    // Name / TemplateId fields stay zero/empty until S114 wires
    // additional AttrCollection ids (NAME=1, ID=10) into the decoder.
    public ImmutableDictionary<long, MonsterData> MonsterDataMap { get; init; }
        = ImmutableDictionary<long, MonsterData>.Empty;

    // ── Window ──
    public WindowRect? WindowRect { get; init; }
    public int WindowWidth { get; init; }
    public int WindowHeight { get; init; }

    // ── Boss raid ──
    public bool BossRaidActive { get; init; }
    public int BossRaidPhase { get; init; }
    public string BossRaidPhaseName { get; init; } = string.Empty;
    public double BossEnrageRemaining { get; init; }
    public string BossTimerText { get; init; } = string.Empty;
    public int BossTotalDamage { get; init; }
    public int BossDps { get; init; }
    public double BossHpEstPct { get; init; } = 1.0;
    public int BossCurrentHp { get; init; }
    public int BossTotalHp { get; init; }
    public BossHpSource BossHpSource { get; init; } = BossHpSource.None;
    // S120 — wall-clock timestamp of the last BossHpEvent application.
    // Used by the MonsterDataMap → Boss* projector to gate "is the packet
    // path still fresh?". When packet-staleness exceeds the projector's
    // window, the projector takes over and overwrites Boss* from the
    // monster table; once a fresh BossHpEvent arrives, it wins again.
    public double BossHpLastPacketSeconds { get; init; }
    public bool BossShieldActive { get; init; }
    public double BossShieldPct { get; init; }
    public int BossBreakingStage { get; init; } = -1;
    public double BossExtinctionPct { get; init; }
    public bool BossInOverdrive { get; init; }
    public bool BossInvincible { get; init; }

    // ── Capture metadata ──
    public double CaptureTimestamp { get; init; }
    public bool RecognitionOk { get; init; }
    public bool PacketActive { get; init; }
    public string ErrorMsg { get; init; } = string.Empty;
    public int IdentityAlertSerial { get; init; }
    public string IdentityAlertTitle { get; init; } = string.Empty;
    public string IdentityAlertMessage { get; init; } = string.Empty;

    public string LevelText => LevelExtra > 0 ? $"{LevelBase}(+{LevelExtra})" : LevelBase.ToString();

    public string HpText => $"{HpCurrent}/{HpMax}";

    public string StaminaText
    {
        get
        {
            var pct = Math.Clamp(StaminaPct, 0.0, 1.0);
            return $"{(int)Math.Round(pct * 100.0)}%";
        }
    }
}

public sealed record SkillSlot
{
    public int Index { get; init; }
    public RectI? Rect { get; init; }
    public SkillSlotState State { get; init; } = SkillSlotState.Unknown;
    public double CooldownPct { get; init; } = 1.0;
    public bool InsufficientEnergy { get; init; }
    public bool Active { get; init; }
    public bool ReadyEdge { get; init; }
    public int ChargeCount { get; init; }
    public int RemainingMs { get; init; }
}

public sealed record BuffEntry
{
    public int Id { get; init; }
    public long Uuid { get; init; }
    public long BeginMs { get; init; }
    public int DurationMs { get; init; }
    public int Layer { get; init; }
    public int Count { get; init; }
    public string Name { get; init; } = string.Empty;
}

public readonly record struct RectI(int X, int Y, int W, int H);

public readonly record struct WindowRect(int Left, int Top, int Right, int Bottom);

public enum SkillSlotState
{
    Unknown = 0,
    Ready = 1,
    Cooldown = 2,
    InsufficientEnergy = 3,
    Active = 4,
}

public enum BossHpSource
{
    None = 0,
    Packet = 1,
    Estimate = 2,
    // S120 — projected from the MonsterDataMap (selected by aggro on
    // SelfUuid, fallback to highest MaxHp Entmonster). Lower precedence
    // than Packet within the freshness window; takes over when the
    // packet path goes stale.
    MonsterData = 3,
}

/// <summary>
/// One row of the near-entity table populated from SyncNearEntities
/// appearances. Holds the entity's wire-format type + the timestamp
/// of first observation, plus the S110 core monster attrs (CurHp,
/// MaxHp, Level) decoded from <c>Entity.Attrs</c> at appearance time.
/// Richer per-entity attributes (TempAttrs, buff list, breaking-stage)
/// land in a follow-up session when the deeper SyncNearDeltaInfo
/// destructuring is ported.
/// </summary>
public readonly record struct EntityTableEntry(
    int EntityType,
    double FirstSeenSeconds,
    int CurHp = 0,
    int MaxHp = 0,
    int Level = 0);

/// <summary>
/// S113 — skeleton port of Python's <c>MonsterData</c> (packet_parser.py
/// 803). Carries the per-monster fields the entity table can't hold:
/// the public-facing identity (Name / TemplateId), the break gauge
/// (BreakingStage / Extinction / MaxExtinction), the shield (
/// ShieldTotal / ShieldMaxTotal / ShieldActive), and the death flag.
/// HP fields mirror the entity table so a delta with HP=0 (HasCurHp=true)
/// can flip IsDead on this row in the same dispatch.
/// Defaults follow Python: BreakingStage=-1 ("not received"), all other
/// numeric/bool slots zero/false. The Name / TemplateId fields stay at
/// their defaults until S114 extends the AttrCollection unpack with
/// NAME (id=1) + ID (id=10).
/// </summary>
public sealed record MonsterData
{
    public long Uuid { get; init; }
    public long Uid { get; init; }
    public string Name { get; init; } = string.Empty;
    public int TemplateId { get; init; }
    public int Hp { get; init; }
    public int MaxHp { get; init; }
    public int Level { get; init; }
    public int BreakingStage { get; init; } = -1;
    public int Extinction { get; init; }
    public int MaxExtinction { get; init; }
    public int Stunned { get; init; }
    public int MaxStunned { get; init; }
    public bool InOverdrive { get; init; }
    public bool ShieldActive { get; init; }
    public int ShieldTotal { get; init; }
    public int ShieldMaxTotal { get; init; }
    public bool IsDead { get; init; }
    // S117 — extended monster flags / attrs.
    public bool IsLockStunned { get; init; }
    public bool StopBreakingTicking { get; init; }
    public int State { get; init; }
    public int DeadType { get; init; }
    public int DeadTime { get; init; }
    public bool FirstAttack { get; init; }
    public long HatedCharId { get; init; }
    public string HatedCharName { get; init; } = string.Empty;
    // S123 — sustained buff list mirrored from
    // AoiSyncDelta.BuffInfos (BuffInfoSync). Authoritative for the
    // monster's currently-active buffs; periodic resync replaces
    // the row wholesale (matches Python's
    // `self._monsters[uuid].buff_list = buffs`). Empty by default
    // so legacy callers (entries created before S123) see no
    // BuffList until a real sync arrives.
    public ImmutableArray<BuffEntry> BuffList { get; init; } = ImmutableArray<BuffEntry>.Empty;
    public double LastUpdateSeconds { get; init; }
}
